store/sqlstore/channel_store_shared_channels.go

```diff
@@ -0,0 +1,275 @@
+		return nil, err
+	}
+
+	if err := s.GetMaster().Insert(sc); err != nil {
+		return nil, errors.Wrapf(err, "save_shared_channel: id=%s", sc.ChannelId)
+	}
+	return sc, nil
+}
+
+// GetSharedChannel fetches a shared channel by channel_id.
+func (s SqlChannelStore) GetSharedChannel(channelId string) (*model.SharedChannel, error) {
+	var sc model.SharedChannel
+
+	query := s.getQueryBuilder().
+		Select("*").
+		From("SharedChannels").
+		Where(sq.Eq{"SharedChannels.ChannelId": channelId})
+
+	squery, args, err := query.ToSql()
+	if err != nil {
+		return nil, errors.Wrapf(err, "getsharedchannel_tosql")
+	}
+
+	if err := s.GetReplica().SelectOne(&sc, squery, args...); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, store.NewErrNotFound("SharedChannel", channelId)
+		}
+		return nil, errors.Wrapf(err, "failed to find channel with ChannelId=%s", channelId)
+	}
+	return &sc, nil
+}
+
+// GetSharedChannels fetches a paginated list of shared channels filtered by SharedChannelSearchOpts.
+func (s SqlChannelStore) GetSharedChannels(offset, limit int, opts store.SharedChannelFilterOpts) ([]*model.SharedChannel, error) {
+	if opts.ExcludeHome && opts.ExcludeRemote {
+		return nil, errors.New("cannot exclude home and remote shared channels")
+	}
+
+	query := s.getSharedChannelsQuery(opts, false)
+	query = query.OrderBy("sc.ShareDisplayName, sc.ShareName").Limit(uint64(limit)).Offset(uint64(offset))
+
+	squery, args, err := query.ToSql()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create query")
+	}
+
+	var channels []*model.SharedChannel
+	_, err = s.GetReplica().Select(&channels, squery, args...)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get channels")
+	}
+	return channels, nil
+}
+
+// GetSharedChannelsCount returns the number of shared channels that would be fetched using SharedChannelSearchOpts.
+func (s SqlChannelStore) GetSharedChannelsCount(opts store.SharedChannelFilterOpts) (int64, error) {
+	if opts.ExcludeHome && opts.ExcludeRemote {
+		return 0, errors.New("cannot exclude home and remote shared channels")
+	}
+
+	query := s.getSharedChannelsQuery(opts, true)
+	squery, args, err := query.ToSql()
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to create query")
+	}
+
+	count, err := s.GetReplica().SelectInt(squery, args...)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to count channels")
+	}
+	return count, nil
+}
+
+func (s SqlChannelStore) getSharedChannelsQuery(opts store.SharedChannelFilterOpts, forCount bool) sq.SelectBuilder {
+	var selectStr string
+	if forCount {
+		selectStr = "count(sc.ChannelId)"
+	} else {
+		selectStr = "sc.*"
+	}
+
+	query := s.getQueryBuilder().
+		Select(selectStr).
+		From("SharedChannels AS sc")
+
+	if opts.TeamId != "" {
+		query = query.Where(sq.Eq{"sc.TeamId": opts.TeamId})
+	}
+
+	if opts.Token != "" {
+		query = query.Where(sq.Eq{"sc.Token": opts.Token})
+	}
+
+	if !opts.ExcludeHome && !opts.ExcludeRemote {
+		return query
+	}
```

Comment:
Is the early return really needed?  It feels like it's not adding that much value.

Reply:
I thought it might help self documenting what combinations make sense, but that obviously didn't work.  It's removed.  https://github.com/mattermost/mattermost-server/pull/16065/commits/5569ffc8fa01fae588f924018afec1743cec6736